model/userdata.go

```diff
@@ -0,0 +1,147 @@
+package model
+
+import (
+	"github.com/coreos/coreos-cloudinit/config/validate"
+	"github.com/kubernetes-incubator/kube-aws/filereader/texttemplate"
+	"github.com/kubernetes-incubator/kube-aws/gzipcompressor"
+
+	"bytes"
+	"encoding/base64"
+	"fmt"
+	"strings"
+	"text/template"
+)
+
+// UserDataPartType defines type of the content in a given part
+// it mostly affects how part should be validated
+type UserDataPartType int
+
+const (
+	USERDATA_SHELL UserDataPartType = iota
+	USERDATA_COREOS_CLOUDINIT
+	USERDATA_YAML
+)
+
+// UserData represents userdata which might be split across multiple storage types
+type UserData struct {
+	Parts map[string]UserDataPart
+}
+
+type UserDataPart struct {
+	Type     UserDataPartType
+	Asset    Asset
+	tmpl     *template.Template
+	tmplData interface{}
+}
```

Comment:
My preference is make `UserDataPart` an interface rather than a struct and implement the interface as `UserDataS3Part` and `UserDataInstancePart` for respective `UserDataPartType`, so that the UserDataPartType and the `switch` block in the `validate` func can be completely omitted(=less cyclomatic complexity)